src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.BoxedTypes.cs

```diff
@@ -89,6 +89,19 @@ public MethodDesc GetSpecialUnboxingThunk(MethodDesc targetMethod, ModuleDesc ow
             return thunk;
         }
 
+        public MethodDesc GetUnspecialUnboxingThunk(MethodDesc targetMethod, ModuleDesc ownerModuleOfThunk)
+        {
+            TypeDesc owningType = targetMethod.OwningType;
+            Debug.Assert(owningType.IsValueType);
+
+            var owningTypeDefinition = (MetadataType)owningType.GetTypeDefinition();
+
+            // Get a reference type that has the same layout as the boxed valuetype.
+            var typeKey = new BoxedValuetypeHashtableKey(owningTypeDefinition, ownerModuleOfThunk);
+            BoxedValueType boxedTypeDefinition = _boxedValuetypeHashtable.GetOrCreateValue(typeKey);
+            return new UnboxingThunk(boxedTypeDefinition, targetMethod);
```

Comment:
This needs to do the hashtable dance for allocation. We use object equality within the type system everywhere to compare two type system objects. This API would make it possible to have two `UnboxingThunk` instances that are the same thing.